### PR TITLE
Fixes to ValuePropagation

### PR DESCRIPTION
**Pull request by liqunl**
This commit contains the following changes
1) Early return in VPHandlers if addKnownObjectConstraints has created
   a constraint. This change will save compilation time in each pass
   of VP.
2) Create KnownObject constraints for newly folded references so that
   the folding can benefit current pass of VP.

Signed-off-by: liqunl <liqunl@ca.ibm.com>

### CODE REVIEW

**Comment by liqunl** [comment]
@andrewcraik Can you review my change?

File: compiler/optimizer/VPHandlers.cpp [hunk 1]
```diff
@@ -1419,15 +1419,15 @@ static const char *getFieldSignature(OMR::ValuePropagation *vp, TR::Node *node,
    return NULL;
    }
 
-static void addKnownObjectConstraints(OMR::ValuePropagation *vp, TR::Node *node)
```
Comment (mgaudet) [inline compiler/optimizer/VPHandlers.cpp]: Maybe a comment here to indicate what true and false mean to consumers would be good?

**Comment by andrewcraik** [comment]
LGTM - over @vijaysun-omr since this is in his area of expertise. @jdmpapin FYI - review not required I don't think, but feel free to comment since you know this code well.

File: compiler/optimizer/VPHandlers.cpp [hunk 2]
```diff
@@ -1516,9 +1525,7 @@ TR::Node *constrainAload(OMR::ValuePropagation *vp, TR::Node *node)
       return node;
 
    // before undertaking aload specific handling see if the default load transformation helps
-   // But even if it does help, node may still be an aload that we can constrain.
-   if (TR::TransformUtil::transformDirectLoad(vp->comp(), node)
-       && node->getOpCodeValue() != TR::aload)
+   if (TR::TransformUtil::transformDirectLoad(vp->comp(), node))
```
Comment (vijaysun-omr) [inline compiler/optimizer/VPHandlers.cpp]: Why did we drop the != TR::aload test ?

Comment (liqunl) [reply compiler/optimizer/VPHandlers.cpp]: If `transformDirectLoad` has done something, it must have changed the node to a constant. For non-collected reference, it'll be an `aconst` node; for collected reference, it'll be a known object index on the symref of the `TR::aload` node. Both cases can be handled by `constrainNewlyFoldedConst` now. Notice this commit contains a change to `constrainNewlyFoldedConst` to create known object constraint.

Comment (vijaysun-omr) [reply compiler/optimizer/VPHandlers.cpp]: Okay so it is specifically to allow the known object constraint to be created in the "current" pass then.

File: compiler/optimizer/VPHandlers.cpp [hunk 3]
```diff
@@ -1540,7 +1547,8 @@ TR::Node *constrainAload(OMR::ValuePropagation *vp, TR::Node *node)
          vp->addGlobalConstraint(node, TR::VPObjectLocation::create(vp, TR::VPObjectLocation::J9ClassObject));
          }
 
-      addKnownObjectConstraints(vp, node);
+      if (addKnownObjectConstraint(vp, node))
```
Comment (keithc-ca) [inline compiler/optimizer/VPHandlers.cpp]: Should this be a call to `addKnownObjectConstraints()` (note the name is plural)?

Comment (liqunl) [reply compiler/optimizer/VPHandlers.cpp]: My bad, I forgot that I have made this little change after tested my branch. I've opened https://github.com/eclipse/omr/pull/1978 for it. @vijaysun-omr Can you review and merge it? I have tested it this time, and so sorry for the trouble.

**Pull request merged**